### PR TITLE
web: structure: centralize assets

### PR DESCRIPTION
**Pull request by aoldershaw**
# Why do we need this PR?

This PR is another refactoring to help with implementing #942. Currently, image assets are scattered around the codebase as hardcoded paths. This makes it difficult to e.g. add a base path to all of the asset URLs.

Prior work: #5301 

# Changes proposed in this pull request

* Introduce `Asset` custom type containing all image assets used in the Elm code
* Add helpers (`Assets.toString`, `Assets.backgroundImage`) for converting the assets to URLs

# Contributor Checklist
- [x] Unit tests

# Reviewer Checklist
- [ ] Code reviewed
- [ ] Tests reviewed

### CODE REVIEW

File: web/elm/benchmarks/Benchmarks.elm [hunk 1]
```diff
@@ -278,7 +279,7 @@ viewBuildHeader session model build =
                     <|
                         [ Icon.icon
                             { sizePx = 40
-                            , image = "ic-add-circle-outline-white.svg"
+                            , image = Assets.AddCircleIcon |> Assets.CircleOutlineIcon
```
Comment (jomsie) [inline web/elm/benchmarks/Benchmarks.elm]: note: i'm kind of in love with this

**Comment by jomsie** [comment]
I'll continue reviewin' this on FF (tomorrow) :)

**Comment by aoldershaw** [comment]
> I'll continue reviewin' this on FF (tomorrow) :)

It's certainly not the most fun PR to review!

File: web/elm/src/Assets.elm [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        SuccessCheckIcon ->
+            basePath ++ [ "ic-success-check.svg" ]
+
+        FailureTimesIcon ->
+            basePath ++ [ "ic-failure-times.svg" ]
+
+        ExclamationTriangleIcon ->
+            basePath ++ [ "ic-exclamation-triangle.svg" ]
+
+        PipelineStatusIcon status ->
```
Comment (jomsie) [inline web/elm/src/Assets.elm]: ❤️

File: web/elm/src/Build/Build.elm [hunk 3]
```diff
@@ -13,6 +13,7 @@ module Build.Build exposing
     )
 
 import Api.Endpoints as Endpoints
+import Assets
```
[the remark this answers is not in the document]
Comment (aoldershaw) [reply web/elm/src/Build/Build.elm]: Huh, weird - I usually have elm-format triggering on every save but I guess it didn't run here!

File: web/elm/tests/TopBarTests.elm [hunk 4]
```diff
@@ -716,7 +722,7 @@ all =
                     Query.find [ id "search-clear" ]
                         >> Query.has
                             [ style "background-image"
-                                "url('public/images/ic-close-white-24px.svg')"
+                                "url(/public/images/ic-close-white-24px.svg)"
```
[the remark this answers is not in the document]
Comment (aoldershaw) [reply web/elm/tests/TopBarTests.elm]: Good catch!

File: web/elm/src/Assets.elm [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                        Windows ->
+                            "windows"
+
+                        Linux ->
+                            "linux"
+            in
+            basePath ++ [ imageName ++ "-logo.svg" ]
+
+        ChevronLeft ->
+            basePath ++ [ "baseline-chevron-left-24px.svg" ]
```
[the remark this answers is not in the document]
Comment (aoldershaw) [reply web/elm/src/Assets.elm]: Yeah, it does feel weird to include the `-24px` suffix in the filename, especially given that the assets aren't always even rendered at 24px!

Initially, I didn't want to update the assets in this PR (it's already spanning 38 files!), but I guess now's as good a time as any!

[the remark this answers is not in the document]
Comment (aoldershaw) [reply web/elm/src/Assets.elm]: They have been renamed. Additionally, I got rid of a bunch of legacy assets that are no longer used

**Comment by jomsie** [comment]
Looks like yarn test is failing... @aoldershaw

**Comment by aoldershaw** [comment]
@jomsie whoops, forgot to rerun the tests after rebasing

**Comment by jomsie** [comment]
i'll continue reviewing after lunch :D

**Pull request merged**